go-apps/meep-platform-ctrl/server/user_authentication.go

```diff
@@ -25,19 +25,206 @@
+	gitlabOauthConfig := &oauth2.Config{
+		ClientID:     gitlabClientId,
+		ClientSecret: gitlabSecret,
+		RedirectURL:  gitlabRedirectUri,
+		Scopes:       []string{"read_user"},
+		Endpoint:     gitlaboauth.Endpoint,
+	}
+	pfmCtrl.oauthConfigs[OAUTH_PROVIDER_GITLAB] = gitlabOauthConfig
+}
+
+// Generate a random state string
+func generateState(n int) (string, error) {
+	data := make([]byte, n)
+	if _, err := io.ReadFull(rand.Reader, data); err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(data), nil
+}
+
+func getUniqueState() (state string, err error) {
+	for i := 0; i < 3; i++ {
+		// Get random state
+		randState, err := generateState(20)
+		if err != nil {
+			log.Error(err.Error())
+			return "", err
+		}
+
+		// Make sure state is unique
+		if _, found := pfmCtrl.loginRequests[randState]; !found {
+			return randState, nil
+		}
+	}
+	return "", errors.New("Failed to generate a random state string")
+}
+
+func uaLoginOAuth(w http.ResponseWriter, r *http.Request) {
+	log.Info("----- OAUTH LOGIN -----")
+
+	// Retrieve query parameters
+	query := r.URL.Query()
+	provider := query.Get("provider")
+
+	// Get provider-specific OAuth config
+	config, found := pfmCtrl.oauthConfigs[provider]
+	if !found {
+		log.Error("Provider config not found for: ", provider)
+		http.Redirect(w, r, pfmCtrl.uri, http.StatusFound)
+		return
+	}
+
+	// Generate unique random state string
+	state, err := getUniqueState()
+	if err != nil {
+		log.Error(err.Error())
+		http.Redirect(w, r, pfmCtrl.uri, http.StatusFound)
+		return
+	}
+
+	// Track oauth request
+	request := &LoginRequest{
+		provider:  provider,
+		timestamp: time.Now(),
+	}
+	pfmCtrl.loginRequests[state] = request
+
+	// Generate provider-specific oauth redirect
+	uri := config.AuthCodeURL(state, oauth2.AccessTypeOnline)
+	http.Redirect(w, r, uri, http.StatusFound)
+}
+
+func uaAuthorize(w http.ResponseWriter, r *http.Request) {
+
+	// Retrieve query parameters
+	query := r.URL.Query()
+	code := query.Get("code")
+	state := query.Get("state")
+
+	// Validate request state
+	request, found := pfmCtrl.loginRequests[state]
+	if !found {
+		log.Error("Login request not found with provided state: ", state)
+		http.Redirect(w, r, pfmCtrl.uri, http.StatusFound)
+		return
+	}
+
+	// Get provider-specific OAuth config
+	config := pfmCtrl.oauthConfigs[request.provider]
+
+	// Retrieve access token
+	token, err := config.Exchange(context.Background(), code)
+	if err != nil {
+		log.Error(err.Error())
+		http.Redirect(w, r, pfmCtrl.uri, http.StatusFound)
+		return
+	}
+
+	// Retrieve User ID
+	var userId string
+	switch request.provider {
+	case OAUTH_PROVIDER_GITHUB:
+		oauthClient := config.Client(context.Background(), token)
+		client := github.NewClient(oauthClient)
+		if client == nil {
+			err = errors.New("Failed to create new GitHub client")
+			log.Error(err.Error())
+			http.Redirect(w, r, pfmCtrl.uri, http.StatusFound)
```

Comment:
Consider if `http.Unauthorized` (401) would be more appropriate here

Reply:
WRT returning 401, we should not use this code on a redirect. If we return a 401 error without the redirect, then the browser shows a blank page with the error message and we must hit the back button in the browser to return to our application.

NOTE: I made some changes in an upcoming PR to return an error string in the redirect URL. This is similar to the error reporting mechanism used by GitHub during OAuth redirects. We could also augment this to open an error dialog on login failure with the error message if we want a more verbose error reporting.

Conclusion: I will report errors via a URL param within the context of a successful redirect for now. This will be pushed soon in another PR.

Reply:
OK, agreed